README.md
<p align=center>
  <a href="https://github.com/TimoKunze/LabelControls/releases/tag/1.2.5">
    <img alt="Release 1.2.5 Release" src="https://img.shields.io/badge/release-1.2.5-0688CB.svg">
  </a>
  <a href="https://github.com/TimoKunze/LabelControls/releases">
    <img alt="Download LabelControls" src="https://img.shields.io/badge/download-latest-0688CB.svg">
  </a>
  <a href="https://github.com/TimoKunze/LabelControls/blob/master/LICENSE">
    <img alt="License" src="https://img.shields.io/badge/license-MIT-0688CB.svg">
  </a>
  <a href="https://www.paypal.com/xclick/business=TKunze71216%40gmx.de&item_name=LabelControls&no_shipping=1&tax=0&currency_code=EUR">
    <img alt="Donate" src="https://img.shields.io/badge/%24-donate-E44E4A.svg">
  </a>
</p>

# LabelControls
An ActiveX control library for Visual Basic 6 that contains various kinds of labels.

I've developed this ActiveX control library in 2013 and did update it on a regular basis until 2016. Currently I have little interest to maintain this project any longer, but I think the code might be of some use to others.

# Before you make changes
If you make changes to the code and deploy the binary, keep in mind that ActiveX controls are COM components and therefore should stay binary compatible as long as you don't change the COM object's, i.e. the ActiveX control's public class name and GUIDs. Otherwise people using these components are likely to end up in the famous COM hell.

# Dependencies
You will need the [Microsoft Windows 10 SDK](https://developer.microsoft.com/en-us/windows/downloads/windows-10-sdk), ATL, and [WTL 10](https://sourceforge.net/projects/wtl/).

## Required Change to ATL
Some version of ATL have a bug in ```AtlIPersistPropertyBag_Load``` which causes crashes. In the file atlcom.h search for ```AtlIPersistPropertyBag_Load```. Inside the implementation of this function search for ```HRESULT hr = pPropBag->Read(pMap[i].szDesc, &var, pErrorLog);```. Make sure that there is this code before this line: ```var.pdispVal = NULL;```.

CHANGELOG
Version 1.2.5 - 30/01/2018

- Upgraded to WTL 10.
- Integrated the GitHub repository into the documentation and the About dialog.



Version 1.2.4 - 11/01/2018

- Project published on GitHub.



Version 1.2.3 - 04/02/2017

- Fixed storing the WindowlessLabel control's TextTruncationStyle property.



Version 1.2.2 - 30/07/2016

- Fixed some memory leaks that occurred when firing certain events.



Version 1.2.1 - 29/01/2016

- FIX: The controls could crash during OLE drag'n'drop.
- FIX: The controls could crash the other process during OLE drag'n'drop, especially when IOLEDataObject::SetDropDescription() had been called.



Version 1.2.0 - 23/01/2016

- Added the FinishOLEDragDrop method to each control. It should not be used if not really needed. It can be used to hide the drag image if execution of the OLEDragDrop event takes a long time or if a message is displayed from within this event.
- Initial support for Windows 10.



Version 1.1.0 - 06/07/2014

- Added the XClick, and XDblClick events for support of the extended mouse buttons.
- Extended the Mouse* events to support the extended mouse buttons.
- FIX: The Font and MouseIcon properties could leak GDI resources.
- Initial support for Windows 8.1.



Version 1.0.5 - 17/10/2013

- Reimplemented the fix for the drawing issues of the WindowlessLabel control. The previous version had unwanted side effects.
- Reimplemented the MouseLeave event of the WindowlessLabel control. Now the control doesn't capture the mouse anymore, so that mouse events are dispatched to the correct control even in complex scenarios where the WindowlessLabel control is partially covered by another windowless control.
- FIX: The IWindowlessLabel::MousePointer property did not work.
- Reactivated data-binding capabilities. When the library is registered, it will call regtlib.exe msdatsrc.tlb before registering itself. This seems to fix the mess caused by the broken Windows Update from the beginning of this year.



Version 1.0.4 - 06/10/2013

- FIX: The WindowlessLabel control crashed, if the AutoSize property has been set to 1 or 2.
- Fixed drawing issues of the WindowlessLabel control.
- Slightly reduced GDI resource usage.



Version 1.0.3 - 13/07/2013

- FIX: A UserControl containing the WindowedLabel control could not be selected in the IDE by clicking on it, if the contained WindowedLabel control has been hit.



Version 1.0.2 - 29/03/2013

- Disabled data-binding capabilities as they cause "Object library not registered" errors when adding the control library to the project. This problem might be related to a recent Windows Update.



Version 1.0.1 - 09/03/2013

- Fixed drawing issues with the WindowlessLabel control for right-to-left layouts.



Version 1.0.0 - 27/01/2013

- First public release.

src/DispIDs.h
// DispIDs.h: Defines a DISPID for each COM property and method we're providing.

// ISysLink
// properties
#define DISPID_SL_APPEARANCE																	1
#define DISPID_SL_AUTOMATICALLYMARKLINKSASVISITED							2
#define DISPID_SL_BACKCOLOR																		3
#define DISPID_SL_BACKSTYLE																		4
#define DISPID_SL_BORDERSTYLE																	5
#define DISPID_SL_CARETLINK																		6
#define DISPID_SL_DETECTDOUBLECLICKS													7
#define DISPID_SL_DISABLEDEVENTS															8
#define DISPID_SL_DONTREDRAW																	9
#define DISPID_SL_ENABLED																			DISPID_ENABLED
#define DISPID_SL_FONT																				10
#define DISPID_SL_FORECOLOR																		11
#define DISPID_SL_HALIGNMENT																	12
#define DISPID_SL_HOTTRACKING																	13
#define DISPID_SL_HOVERTIME																		14
#define DISPID_SL_HWND																				15
#define DISPID_SL_HWNDTOOLTIP																	16
#define DISPID_SL_LINKS																				17
#define DISPID_SL_MOUSEICON																		18
#define DISPID_SL_MOUSEPOINTER																19
#define DISPID_SL_PROCESSCONTEXTMENUKEYS											20
#define DISPID_SL_PROCESSRETURNKEY														21
#define DISPID_SL_REGISTERFOROLEDRAGDROP											22
#define DISPID_SL_RIGHTTOLEFT																	23
#define DISPID_SL_SHOWTOOLTIPS																24
#define DISPID_SL_SUPPORTOLEDRAGIMAGES												25
#define DISPID_SL_TEXT																				DISPID_VALUE
#define DISPID_SL_USEMNEMONIC																	26
#define DISPID_SL_USESYSTEMFONT																27
#define DISPID_SL_USEVISUALSTYLE															28
#define DISPID_SL_VERSION																			29
// fingerprint
#define DISPID_SL_APPID																				500
#define DISPID_SL_APPNAME																			501
#define DISPID_SL_APPSHORTNAME																502
#define DISPID_SL_BUILD																				503
#define DISPID_SL_CHARSET																			504
#define DISPID_SL_ISRELEASE																		505
#define DISPID_SL_PROGRAMMER																	506
#define DISPID_SL_TESTER																			507
// methods
#define DISPID_SL_ABOUT																				DISPID_ABOUTBOX
#define DISPID_SL_GETIDEALSIZE																30
#define DISPID_SL_HITTEST																			31
#define DISPID_SL_LOADSETTINGSFROMFILE												32
#define DISPID_SL_REFRESH																			DISPID_REFRESH
#define DISPID_SL_SAVESETTINGSTOFILE													33
#define DISPID_SL_FINISHOLEDRAGDROP														34


// _ISysLinkEvents
// methods
#define DISPID_SLE_CARETCHANGED																1
#define DISPID_SLE_CLICK																			2
#define DISPID_SLE_CONTEXTMENU																3
#define DISPID_SLE_CUSTOMDRAW																	4
#define DISPID_SLE_CUSTOMIZETEXTDRAWING												5
#define DISPID_SLE_DBLCLICK																		6
#define DISPID_SLE_DESTROYEDCONTROLWINDOW											7
#define DISPID_SLE_KEYDOWN																		8
#define DISPID_SLE_KEYPRESS																		9
#define DISPID_SLE_KEYUP																			10
#define DISPID_SLE_LINKGETINFOTIPTEXT													11
#define DISPID_SLE_LINKMOUSEENTER															12
#define DISPID_SLE_LINKMOUSELEAVE															13
#define DISPID_SLE_MCLICK																			14
#define DISPID_SLE_MDBLCLICK																	15
#define DISPID_SLE_MOUSEDOWN																	16
#define DISPID_SLE_MOUSEENTER																	17
#define DISPID_SLE_MOUSEHOVER																	18
#define DISPID_SLE_MOUSELEAVE																	19
#define DISPID_SLE_MOUSEMOVE																	20
#define DISPID_SLE_MOUSEUP																		21
#define DISPID_SLE_OLEDRAGDROP																22
#define DISPID_SLE_OLEDRAGENTER																23
#define DISPID_SLE_OLEDRAGLEAVE																24
#define DISPID_SLE_OLEDRAGMOUSEMOVE														25
#define DISPID_SLE_OPENLINK																		DISPID_VALUE
#define DISPID_SLE_RCLICK																			26
#define DISPID_SLE_RDBLCLICK																	27
#define DISPID_SLE_RECREATEDCONTROLWINDOW											28
#define DISPID_SLE_RESIZEDCONTROLWINDOW												29
#define DISPID_SLE_TEXTCHANGED																30
#define DISPID_SLE_XCLICK																			31
#define DISPID_SLE_XDBLCLICK																	32


// ILink
// properties
#define DISPID_L_CARET																				1
#define DISPID_L_DRAWASNORMALTEXT															2
#define DISPID_L_ENABLED																			3
#define DISPID_L_HOT																					4
#define DISPID_L_ID																						5
#define DISPID_L_INDEX																				6
#define DISPID_L_TEXT																					7
#define DISPID_L_URL																					DISPID_VALUE
#define DISPID_L_VISITED																			8
// methods


// ILinks
// properties
#define DISPID_LS_CASESENSITIVEFILTERS												1
#define DISPID_LS_COMPARISONFUNCTION													2
#define DISPID_LS_FILTER																			3
#define DISPID_LS_FILTERTYPE																	4
#define DISPID_LS_ITEM																				DISPID_VALUE
#define DISPID_LS__NEWENUM																		DISPID_NEWENUM
// methods
#define DISPID_LS_CONTAINS																		5
#define DISPID_LS_COUNT																				6


// IWindowedLabel
// properties
#define DISPID_WLBL_APPEARANCE																1
#define DISPID_WLBL_AUTOSIZE																	2
#define DISPID_WLBL_BACKCOLOR																	3
#define DISPID_WLBL_BACKSTYLE																	4
#define DISPID_WLBL_BORDERSTYLE																5
#define DISPID_WLBL_CLIPLASTLINE															6
#define DISPID_WLBL_DISABLEDEVENTS														7
#define DISPID_WLBL_DONTREDRAW																8
#define DISPID_WLBL_ENABLED																		DISPID_ENABLED
#define DISPID_WLBL_FONT																			9
#define DISPID_WLBL_FORECOLOR																	10
#define DISPID_WLBL_HALIGNMENT																11
#define DISPID_WLBL_HOVERTIME																	12
#define DISPID_WLBL_HWND																			13
#define DISPID_WLBL_MOUSEICON																	14
#define DISPID_WLBL_MOUSEPOINTER															15
#define DISPID_WLBL_OWNERDRAWN																16
#define DISPID_WLBL_REGISTERFOROLEDRAGDROP										17
#define DISPID_WLBL_RIGHTTOLEFT																18
#define DISPID_WLBL_SUPPORTOLEDRAGIMAGES											19
#define DISPID_WLBL_TEXT																			DISPID_VALUE
#define DISPID_WLBL_TEXTTRUNCATIONSTYLE												20
#define DISPID_WLBL_USEMNEMONIC																21
#define DISPID_WLBL_USESYSTEMFONT															22
#define DISPID_WLBL_VERSION																		23
// fingerprint
#define DISPID_WLBL_APPID																			500
#define DISPID_WLBL_APPNAME																		501
#define DISPID_WLBL_APPSHORTNAME															502
#define DISPID_WLBL_BUILD																			503
#define DISPID_WLBL_CHARSET																		504
#define DISPID_WLBL_ISRELEASE																	505
#define DISPID_WLBL_PROGRAMMER																506
#define DISPID_WLBL_TESTER																		507
// methods
#define DISPID_WLBL_ABOUT																			DISPID_ABOUTBOX
#define DISPID_WLBL_LOADSETTINGSFROMFILE											24
#define DISPID_WLBL_REFRESH																		DISPID_REFRESH
#define DISPID_WLBL_SAVESETTINGSTOFILE												25
#define DISPID_WLBL_FINISHOLEDRAGDROP													26


// _IWindowedLabelEvents
// methods
#define DISPID_WLBLE_CLICK																		1
#define DISPID_WLBLE_CONTEXTMENU															2
#define DISPID_WLBLE_DBLCLICK																	3
#define DISPID_WLBLE_DESTROYEDCONTROLWINDOW										4
#define DISPID_WLBLE_MCLICK																		5
#define DISPID_WLBLE_MDBLCLICK																6
#define DISPID_WLBLE_MOUSEDOWN																7
#define DISPID_WLBLE_MOUSEENTER																8
#define DISPID_WLBLE_MOUSEHOVER																9
#define DISPID_WLBLE_MOUSELEAVE																10
#define DISPID_WLBLE_MOUSEMOVE																11
#define DISPID_WLBLE_MOUSEUP																	12
#define DISPID_WLBLE_OLEDRAGDROP															13
#define DISPID_WLBLE_OLEDRAGENTER															14
#define DISPID_WLBLE_OLEDRAGLEAVE															15
#define DISPID_WLBLE_OLEDRAGMOUSEMOVE													16
#define DISPID_WLBLE_OWNERDRAW																17
#define DISPID_WLBLE_RCLICK																		18
#define DISPID_WLBLE_RDBLCLICK																19
#define DISPID_WLBLE_RECREATEDCONTROLWINDOW										20
#define DISPID_WLBLE_RESIZEDCONTROLWINDOW											21
#define DISPID_WLBLE_TEXTCHANGED															DISPID_VALUE
#define DISPID_WLBLE_XCLICK																		22
#define DISPID_WLBLE_XDBLCLICK																23


// IWindowlessLabel
// properties
#define DISPID_WLLBL_APPEARANCE																1
#define DISPID_WLLBL_AUTOSIZE																	2
#define DISPID_WLLBL_BACKCOLOR																3
#define DISPID_WLLBL_BACKSTYLE																4
#define DISPID_WLLBL_BORDERSTYLE															5
#define DISPID_WLLBL_CLIPLASTLINE															6
#define DISPID_WLLBL_DISABLEDEVENTS														7
#define DISPID_WLLBL_DONTREDRAW																8
#define DISPID_WLLBL_ENABLED																	DISPID_ENABLED
#define DISPID_WLLBL_FONT																			9
#define DISPID_WLLBL_FORECOLOR																10
#define DISPID_WLLBL_HALIGNMENT																11
#define DISPID_WLLBL_HOVERTIME																12
#define DISPID_WLLBL_MOUSEICON																13
#define DISPID_WLLBL_MOUSEPOINTER															14
#define DISPID_WLLBL_REGISTERFOROLEDRAGDROP										15
#define DISPID_WLLBL_RIGHTTOLEFT															16
#define DISPID_WLLBL_SUPPORTOLEDRAGIMAGES											17
#define DISPID_WLLBL_TEXT																			DISPID_VALUE
#define DISPID_WLLBL_TEXTTRUNCATIONSTYLE											18
#define DISPID_WLLBL_USEMNEMONIC															19
#define DISPID_WLLBL_USESYSTEMFONT														20
#define DISPID_WLLBL_VALIGNMENT																21
#define DISPID_WLLBL_VERSION																	22
#define DISPID_WLLBL_WORDWRAPPING															23
// fingerprint
#define DISPID_WLLBL_APPID																		500
#define DISPID_WLLBL_APPNAME																	501
#define DISPID_WLLBL_APPSHORTNAME															502
#define DISPID_WLLBL_BUILD																		503
#define DISPID_WLLBL_CHARSET																	504
#define DISPID_WLLBL_ISRELEASE																505
#define DISPID_WLLBL_PROGRAMMER																506
#define DISPID_WLLBL_TESTER																		507
// methods
#define DISPID_WLLBL_ABOUT																		DISPID_ABOUTBOX
#define DISPID_WLLBL_LOADSETTINGSFROMFILE											24
#define DISPID_WLLBL_REFRESH																	DISPID_REFRESH
#define DISPID_WLLBL_SAVESETTINGSTOFILE												25
#define DISPID_WLLBL_FINISHOLEDRAGDROP												26


// _IWindowlessLabelEvents
// methods
#define DISPID_WLLBLE_CLICK																		1
#define DISPID_WLLBLE_CONTEXTMENU															2
#define DISPID_WLLBLE_DBLCLICK																3
#define DISPID_WLLBLE_MCLICK																	4
#define DISPID_WLLBLE_MDBLCLICK																5
#define DISPID_WLLBLE_MOUSEDOWN																6
#define DISPID_WLLBLE_MOUSEENTER															7
#define DISPID_WLLBLE_MOUSEHOVER															8
#define DISPID_WLLBLE_MOUSELEAVE															9
#define DISPID_WLLBLE_MOUSEMOVE																10
#define DISPID_WLLBLE_MOUSEUP																	11
#define DISPID_WLLBLE_OLEDRAGDROP															12
#define DISPID_WLLBLE_OLEDRAGENTER														13
#define DISPID_WLLBLE_OLEDRAGLEAVE														14
#define DISPID_WLLBLE_OLEDRAGMOUSEMOVE												15
#define DISPID_WLLBLE_RCLICK																	16
#define DISPID_WLLBLE_RDBLCLICK																17
#define DISPID_WLLBLE_RESIZEDCONTROLWINDOW										18
#define DISPID_WLLBLE_TEXTCHANGED															DISPID_VALUE
#define DISPID_WLLBLE_XCLICK																	19
#define DISPID_WLLBLE_XDBLCLICK																20


// IOLEDataObject
// properties
// methods
#define DISPID_ODO_CLEAR																			1
#define DISPID_ODO_GETCANONICALFORMAT													2
#define DISPID_ODO_GETDATA																		3
#define DISPID_ODO_GETFORMAT																	4
#define DISPID_ODO_SETDATA																		5
#define DISPID_ODO_GETDROPDESCRIPTION													6
#define DISPID_ODO_SETDROPDESCRIPTION													7
